### PR TITLE
Data type browser

### PR DESCRIPTION
**Pull request by ApurvaSinha17**
Added code for Data type browser

### CODE REVIEW

**Comment by zednis** [comment]
Let's complete pull request #2 and then merge the new master into this branch so the diffs in this pull request are only related to the datatype browser.

edit - I'll talk with @mrpatrickwest to make sure we close out #2 soon.

File: opensearch/descriptions/datatypes.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	xmlns:time="http://a9.com/-/opensearch/extensions/time/1.0/"
+	xmlns:dco="http://data.deepcarbon.net/dco-s2s/ontologies/dco-s2s.ttl#"
+	xmlns:core="http://escience.rpi.edu/ontology/sesf/s2s-core/4/0/">
+
+  <ShortName>DCO Data Portal Datatype Search</ShortName>
+  <Description>DCO Data Portal Datatype Search Service. This opensearch
+  document describes how to call the various RESTful services in order
+  to fill s2s facets (select creation year of the data type, creator of the data type, parameters of the data type, source standard of the data type, subject area of the data type, and the result set.)</Description>
+  <Tags></Tags>
+  <Contact>Apurva Kumar Sinha</Contact>
```
Comment (zednis) [inline opensearch/descriptions/datatypes.xml]: change contact to web@mail.deepcarbon.net

File: opensearch/descriptions/datatypes.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  <Url type="application/json" rel="dco:SourceStandardQuery" template="http://data.deepcarbon.net/dco-s2s/opensearch/services/datatypes/search.php?request=sourcestandard&amp;creator={dco:Creator?}&amp;parameter={dco:Parameter?}&amp;creationyear={dco:CreatedAtTime?}&amp;subjectarea={dco:SubjectArea?}" />
+  <Url type="application/json" rel="dco:SubjectAreaQuery" template="http://data.deepcarbon.net/dco-s2s/opensearch/services/datatypes/search.php?request=subjectarea&amp;creator={dco:Creator?}&amp;parameter={dco:Parameter?}&amp;sourcestandard={dco:SourceStandard?}&amp;creationyear={dco:CreatedAtTime?}" />
+  <Url type="text/html" rel="dco:ResultsListQuery" template="http://data.deepcarbon.net/dco-s2s/opensearch/services/datatypes/search.php?request=datatypes&amp;creator={dco:Creator?}&amp;parameter={dco:Parameter?}&amp;sourcestandard={dco:SourceStandard?}&amp;subjectarea={dco:SubjectArea?}&amp;creationyear={dco:CreatedAtTime?}&amp;limit={count}&amp;offset={startIndex}" />
+
+  <LongName>DCO Data Portal Datatype Search Service</LongName>
+  <Developer>Apurva Kumar Sinha</Developer>
```
Comment (zednis) [inline opensearch/descriptions/datatypes.xml]: Change developer to "DCO Data Science Team"

**Comment by zednis** [comment]
let's use  camelCase on parameter names in the open search service.

"subjectarea"  -> "subjectArea"
"sourcestandard" -> "sourceStandard"
 "creationyear" -> "creationYear"

File: opensearch/services/datatypes/datatypes.php [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+	* @param string $type search type (e.g. 'parameters', 'authors', 'keywords')
+	* @return string query header component (e.g. 'SELECT ?id ?label')
+	*/
+	public function getQueryHeader($type) {
+	
+		$header = "";
+		switch($type) {
+			case "datatypes":
+				$header .= "?datatype ?dco_id ?label ?year ?parameterLabel ?access ";
+				$header .= '(GROUP_CONCAT(DISTINCT ?creator ; SEPARATOR=",") AS ?creator) ';
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php]: I am not sure we should be defining a new variable (i.e. ?creator, ?creator_label, ?sourceStandard, ?sourceStandard_label)  with the same name as the existing variable.

See [project.php](https://github.com/tetherless-world/dco-s2s-browsers/blob/master/opensearch/services/projects/projects.php) as an example.

File: opensearch/descriptions/datatypes.xml (hunk 1 above)
Comment (ApurvaSinha17) [inline opensearch/descriptions/datatypes.xml]: changed

File: opensearch/descriptions/datatypes.xml (hunk 2 above)
Comment (ApurvaSinha17) [inline opensearch/descriptions/datatypes.xml]: changed

File: opensearch/services/datatypes/datatypes.php [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+		
+		$body = "";
+		switch($type) {
+			case "creationYear":
+				$body .= "?datatype a dco:DataType . ";
+				$body .= "?datatype  dco:createdAtTime ?t . ";
+				$body .= "BIND(str(?t) AS ?year) . ";
+				break;
+				
+			case "creator":
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:325]: I am pretty sure this query is broken.  Have you tested it?
1. the creator instance should not have type dco:creator
2. the creator instance should not have the role instance as a type
3. this query should include the ?id and ?label variables (see other comments below)

File: opensearch/services/datatypes/datatypes.php [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+				$body .= "BIND(str(?l) AS ?parameterLabel) . ";
+				break;
+
+			case "sourceStandard":
+				$body .= "?datatype a dco:DataType . ";
+				$body .= "?datatype dco:sourceStandard [rdfs:label ?l] . ";
+				$body .= "BIND(str(?l) AS ?sourceStandard_label) . ";
+				break;
+
+			case "subjectArea":
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:349]: The expected parameters to return for this query should be 

`?id ?label (COUNT(DISTINCT ?datatype) AS ?count)`

and yet the query body does not contain ?id or ?label.

Rewrite to

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype vivo:hasSubjectArea ?id . ";
$body .= "?id skos:prefLabel ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

File: opensearch/services/datatypes/datatypes.php [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+			case "creator":
+				$body .= "?datatype a dco:DataType . ";
+				$body .= "?datatype prov:qualifiedAttribution [prov:hadRole ?r ] . ";
+				$body .= "?r a prov:Role . ";
+				$body .= "?c rdf:type ?r . ";
+				$body .= "?c a dco:creator . ";
+				$body .= "BIND(str(?c) AS ?creator) . ";
+				break;
+
+			case "parameter":
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:333]: The expected parameters to return for this query should be 

`?id ?label (COUNT(DISTINCT ?datatype) AS ?count)`

and yet the query body does not contain ?id or ?label.

Rewrite to

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype dco:hasParameter ?id . ";
$body .= "?id rdfs:label ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

File: opensearch/services/datatypes/datatypes.php [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+	/**
+	  * Return SPARQL query WHERE clause minus constraint clauses for specified search type
+	  * @param string $type search type (e.g. 'datasets', 'authors', 'keywords')
+	  * @return string WHERE clause component minus constraint clauses (e.g. '?dataset a dcat:Dataset . ')
+	  */
+	public function getQueryBody($type) {
+		
+		$body = "";
+		switch($type) {
+			case "creationYear":
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:319]: 1. You should add a case statement to the header for ?year (see comment above)

File: opensearch/services/datatypes/datatypes.php [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+				$header .= "?datatype ?dco_id ?label ?year ?parameterLabel ?access ";
+				$header .= '(GROUP_CONCAT(DISTINCT ?creator ; SEPARATOR=",") AS ?creator) ';
+				$header .= '(GROUP_CONCAT(DISTINCT ?creator_label ; SEPARATOR=",") AS ?creator_label) ';
+				$header .= '(GROUP_CONCAT(DISTINCT ?sourceStandard ; SEPARATOR=",") AS ?sourceStandard) ';
+                $header .= '(GROUP_CONCAT(DISTINCT ?sourceStandard_label ; SEPARATOR=",") AS ?sourceStandard_label) ';
+				break;
+			case "count":
+				$header .= "(count(DISTINCT ?datatype) AS ?count)";
+				break;
+			default:
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:277]: Add a new case statement for when the query parameter is "creationYear".  In this case we are not return ?id and ?label but just ?year.  You will likewise need to add a case for "creationYear" in the getQueryFooter function.

```
case "creationYear":
    $header .= '?year' ;
    break;
```

File: opensearch/services/datatypes/datatypes.php [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+		$footer = "";
+		switch($type) {
+			case "datatypes":
+				$footer .= " GROUP BY ?datatype ?dco_id ?label ?year ?parameterLabel ?access";
+				$footer .= " ORDER BY ?label";
+				if ($limit)	$footer .= " LIMIT $limit OFFSET $offset";
+				break;
+			case "count":
+				break;
+			default:
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php]: Add a case for "creationYear".

```
case "creationYear":
    $footer .= " GROUP BY ?year";
    break;
```

File: opensearch/services/datatypes/datatypes.php (hunk 5 above)
Comment (ApurvaSinha17) [inline opensearch/services/datatypes/datatypes.php:349]: Should it be like this? I checked the SPARQL query using this and the results were coming.

```
$body .= "?datatype a dco:DataType . ";
$body .= "?datatype dco:dataTypeSubjectArea ?id . ";
$body .= "?id a skos:Concept . ";
$body .= "?id rdfs:label ?l .";
$body .= "BIND(str(?l) AS ?label) . ";
```

Comment (zednis) [inline opensearch/services/datatypes/datatypes.php:349]: If we proceed with using dco:dataTypeSubjectArea then that would be the correct query.

edit - are we using skos:prefLabel with the subject area concepts or just rdfs:label?

File: opensearch/services/datatypes/datatypes.php [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+				$body .= "OPTIONAL { ?datatype dco:expectedUses ?use . } ";
+				$body .= "OPTIONAL { ?datatype dco:hasParameter ?param . ?param rdfs:label ?pl . } ";
+				$body .= "OPTIONAL { ?datatype dco:sourceDataType ?s . ?s rdfs:label ?sourceLabel . } ";
+				$body .= "BIND(str(?l) AS ?label) . ";
+				$body .= "BIND(str(?id) AS ?dco_id) . ";
+				$body .= "BIND(str(?ct) AS ?createdTime) . ";
+				$body .= "BIND(str(?mt) AS ?modifiedTime) . ";
+				$body .= "BIND(str(?sl) AS ?sourceStandard_label) . ";
+				$body .= "BIND(str(?use) AS ?use) . ";
+				$body .= "BIND(str(?pl) AS ?parameterLabel) . ";
```
Comment (zednis) [inline opensearch/services/datatypes/datatypes.php]: Can there be more than 1 parameter per datatype?

Comment (ApurvaSinha17) [inline opensearch/services/datatypes/datatypes.php]: Yes, there is one sample data type instance in UDCO that has multiple parameters.

File: opensearch/services/datatypes/datatypes.php (hunk 4 above)
Comment (ApurvaSinha17) [inline opensearch/services/datatypes/datatypes.php:325]: Made corrections

File: opensearch/services/datatypes/datatypes.php (hunk 7 above)
Comment (ApurvaSinha17) [inline opensearch/services/datatypes/datatypes.php:319]: added

**Comment by mrpatrickwest** [comment]
Merged master into this branch, updated to use https, pushed to udco (changed urls there) and it works. Ready to merge the pull request.

**Pull request merged**